### PR TITLE
Prerender webapp for bots using my dev server as a prerender node (WIP)

### PR DESCRIPTION
**Pull request by jgarciao**
This nginx configuration detects bots like googlebot and others (based on user-agent) and serves the content for them using a prerender service, which is like a headless chrome. 

I've tried this solution in my dev server and it works well. You can see the results searching this on google: 
- `site:45.55.104.129`
Shows 310 pages indexed
- By contrast:
`site:flathub.org`
Shows only 3 pages indexed

In my dev server I've changed this nginx conf files to make it work:

- https://github.com/flathub/linux-store-frontend/blob/linux-store-frontend-0.6.x/conf/nginx/nginx.conf
- https://github.com/flathub/linux-store-frontend/blob/linux-store-frontend-0.6.x/conf/nginx/default.conf

This PR tries to adapt the same changes to ansible rules. I'm not familiar with ansible so it's possible there are bugs

I've also setup a prerender node in my dev server and I've configured the firewall to allow queries from flathub servers.

### CODE REVIEW

File: roles/frontend/templates/nginx-default.d-webroot.conf.j2 [hunk 1]
```diff
@@ -1,9 +1,32 @@
...
+
+location @prerender {
+
     proxy_cache frontend_cache;
     proxy_cache_valid 200 302 10m;
     proxy_cache_valid 301      1h;
     proxy_cache_valid any      1m;
-    proxy_pass {{ frontend_nginx_webroot_origin }};
+
+    proxy_buffering on;
```
Comment (ramcq) [inline roles/frontend/templates/nginx-default.d-webroot.conf.j2]: This seems worth having anyway - maybe move it to templates/nginx-conf.d-cache.conf.j2

File: roles/frontend/templates/nginx-default.d-webroot.conf.j2 [hunk 2]
```diff
@@ -1,9 +1,32 @@
...
+    proxy_cache_revalidate on;
```
Comment (ramcq) [inline roles/frontend/templates/nginx-default.d-webroot.conf.j2]: Can be removed, we have it in templates/nginx-conf.d-cache.conf.j2

File: roles/frontend/templates/nginx-default.d-webroot.conf.j2 [hunk 3]
```diff
@@ -1,9 +1,32 @@
...
+    proxy_cache_min_uses 1;
```
Comment (ramcq) [inline roles/frontend/templates/nginx-default.d-webroot.conf.j2]: Is the default, we can remove.

File: roles/frontend/templates/nginx-default.d-webroot.conf.j2 [hunk 4]
```diff
@@ -1,9 +1,32 @@
...
+    proxy_cache_use_stale error timeout updating http_500 http_502 http_503 http_504;
```
Comment (ramcq) [inline roles/frontend/templates/nginx-default.d-webroot.conf.j2]: Can remove, is already present in templates/nginx-conf.d-cache.conf.j2

File: roles/frontend/tasks/nginx.yml [hunk 5]
```diff
@@ -32,6 +32,13 @@
   notify:
    - reload nginx
 
+- name: install nginx conf.d/map.conf
+  copy:
+    src: nginx-map.conf
```
Comment (ramcq) [inline roles/frontend/tasks/nginx.yml]: Should be named nginx-conf.d-map.conf for consistency. I might actually name it prerender-map.conf when it's installed anyway.

File: roles/frontend/defaults/main.yml [hunk 6]
```diff
@@ -2,6 +2,8 @@ frontend_dehydrated_domains:
  - "{{ frontend_nginx_ssl_name }}"
 frontend_nginx_api_servers:
  - apps.flathub.org:8080
+frontend_nginx_prerender_servers:
+ - 45.55.104.129:3000
```
Comment (ramcq) [inline roles/frontend/defaults/main.yml:6]: Is there any real benefit to using your beta server (and hence introducing a production dependency on it) versus using the public service on prerender.io? Or could we use both?

**Review by ramcq — CHANGES_REQUESTED** [review]

File: roles/frontend/templates/nginx-default.d-webroot.conf.j2 (hunk 1 above)
Comment (jgarciao) [reply roles/frontend/templates/nginx-default.d-webroot.conf.j2]: Done

File: roles/frontend/templates/nginx-default.d-webroot.conf.j2 (hunk 2 above)
Comment (jgarciao) [reply roles/frontend/templates/nginx-default.d-webroot.conf.j2]: Done

File: roles/frontend/templates/nginx-default.d-webroot.conf.j2 [hunk 7]
```diff
@@ -1,9 +1,32 @@
 location / {
+    try_files $uri @prerender;
```
Comment (ramcq) [inline roles/frontend/templates/nginx-default.d-webroot.conf.j2:2]: We also don't need this - the @prerender block should be unconditionally used for / - there is no meaningful file root set for the webroot.

File: roles/frontend/templates/nginx-default.d-webroot.conf.j2 (hunk 3 above)
Comment (jgarciao) [reply roles/frontend/templates/nginx-default.d-webroot.conf.j2]: Done

File: roles/frontend/templates/nginx-default.d-webroot.conf.j2 (hunk 4 above)
Comment (jgarciao) [reply roles/frontend/templates/nginx-default.d-webroot.conf.j2]: Done

File: roles/frontend/tasks/nginx.yml (hunk 5 above)
Comment (jgarciao) [reply roles/frontend/tasks/nginx.yml]: Fixed

File: roles/frontend/defaults/main.yml (hunk 6 above)
Comment (jgarciao) [reply roles/frontend/defaults/main.yml:6]: Prerender.io free tier is for 250 pages or less. We already have more
https://prerender.io/pricing

File: roles/frontend/templates/nginx-default.d-webroot.conf.j2 (hunk 7 above)
Comment (jgarciao) [reply roles/frontend/templates/nginx-default.d-webroot.conf.j2:2]: This I don't know. I just copied from a gist and it works fine

**Comment by ramcq** [comment]
Merged a version of this to master.

**Pull request closed without merging**